### PR TITLE
feat: enable pod autoscaling

### PR DESCRIPTION
**Pull request by rokroskar**
Provides an optional HPA template to enable autoscaling, configurable through the values file.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -3,7 +3,6 @@ repos:
     rev: stable
     hooks:
     - id: black
-      language_version: python3.7
```
Comment (ableuler) [inline .pre-commit-config.yaml:6]: Is this change on purpose? Just asking as it seems unrelated to the rest of this PR.

File: helm-chart/renku-notebooks/values.yaml [hunk 2]
```diff
@@ -21,6 +21,13 @@ gitlab:
 
 replicaCount: 1
 
+## Configure autoscaling
+autoscaling:
+  enabled: false
```
Comment (ableuler) [inline helm-chart/renku-notebooks/values.yaml:26]: I'm wondering if we should fully adopt the bitnami convention and have two values files in each chart. A `values.yaml` and a `production-values.yaml` which would have autoscaling on per default and a minimum of two replicas. Or maybe it's enough to do this on the RenkuLab chart level?

File: helm-chart/renku-notebooks/templates/hpa.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+apiVersion: autoscaling/v1
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ template "notebooks.fullname" . }}
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: {{ template "notebooks.fullname" . }}
+  minReplicas: {{ .Values.autoscaling.minReplicas }}
```
Comment (ableuler) [inline helm-chart/renku-notebooks/templates/hpa.yaml:11]: Do you know if the HPA guarantees that the replicas are spread out over at least two nodes?

**Review by ableuler — COMMENTED** [review]
Great and long-needed addition. The main further improvement I see would be to use `autoscaling/v2beta2` instead of `autoscaling/v1` and add autoscaling also based on memory usage. But if you don't want to tackle this right now I'm happy to merge the PR "as-is" as this is already a great improvement over the status quo.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (rokroskar) [reply .pre-commit-config.yaml:6]: the pre-commit hook was failing because I'm using python 3.8

File: helm-chart/renku-notebooks/values.yaml (hunk 2 above)
Comment (rokroskar) [reply helm-chart/renku-notebooks/values.yaml:26]: yes, I think it's enough to do it there

File: helm-chart/renku-notebooks/templates/hpa.yaml (hunk 3 above)
Comment (rokroskar) [reply helm-chart/renku-notebooks/templates/hpa.yaml:11]: pretty sure it just hands the request to the scheduler - you'd need to have anti-affinity set separately. This isn't about HA I guess

**Comment by rokroskar** [comment]
Why would you scale the notebooks service based on memory usage? I think the memory usage should be pretty constant.

File: helm-chart/renku-notebooks/templates/hpa.yaml (hunk 3 above)
Comment (pameladelgado) [reply helm-chart/renku-notebooks/templates/hpa.yaml:11]: I think it does

**Review by pameladelgado — DISMISSED** [review]
I tested this and it works wonderfully.

**Comment by ableuler** [comment]
I agree that I'd expect the CPU usage to scale more linearly with the number of requests per time, but I think that memory based autoscaling can make our services a bit more robust against things like memory leaks, requests piling up because of slow upstream services etc.

**Comment by rokroskar** [comment]
Thanks for testing @pameladelgado - I pushed a change to make some sensible defaults so all that is needed is to set the `autoscaling.enabled=true` instead of having to set the whole block.

**Comment by rokroskar** [comment]
> I agree that I'd expect the CPU usage to scale more linearly with the number of requests per time, but I think that memory based autoscaling can make our services a bit more robust against things like memory leaks, requests piling up because of slow upstream services etc.

Sure, good point - I'd say lets see if this is a possibility/issue and if so add a second metric.

**Comment by pameladelgado** [comment]
> Thanks for testing @pameladelgado - I pushed a change to make some sensible defaults so all that is needed is to set the `autoscaling.enabled=true` instead of having to set the whole block.

Scaling down does take a while, I guess this is normal.
Regarding memory I was going to suggest the same as @ableuler but I also thought that for notebooks memory might not be a critical resource.

**Review by pameladelgado — APPROVED** [review]

File: helm-chart/renku-notebooks/templates/hpa.yaml (hunk 3 above)
Comment (ableuler) [reply helm-chart/renku-notebooks/templates/hpa.yaml:11]: You're right @pameladelgado. The scheduler in its default settings will try to spread out pods, but it won't guarantee anything (it's just entering the big priority scoring function): 
```
SelectorSpreadPriority: Spreads Pods across hosts, considering Pods that belong to the same Service, StatefulSet or ReplicaSet.
```
